File: test-ml-models.js
/**
 * Test ML model access for cascading fallback system
 */

// Load environment variables from .env file
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

function loadEnvFile() {
  try {
    const envPath = path.join(__dirname, '.env');
    const envContent = fs.readFileSync(envPath, 'utf8');

    envContent.split('\n').forEach(line => {
      const [key, value] = line.split('=');
      if (key && value && !key.startsWith('#')) {
        process.env[key.trim()] = value.trim();
      }
    });
  } catch (error) {
    console.log('Warning: Could not load .env file:', error.message);
  }
}

loadEnvFile();

async function testMalaysianSFWClassifier() {
  console.log("=" * 60);
  console.log("TESTING MALAYSIAN SFW CLASSIFIER VIA API");
  console.log("=" * 60);
  
  const HF_API_KEY = process.env.HUGGINGFACE_API_KEY;
  if (!HF_API_KEY) {
    console.log("❌ No Hugging Face API key found");
    return false;
  }
  
  console.log(`✅ Using HF API key: ${HF_API_KEY.substring(0, 10)}...`);
  
  const testTexts = ["bodoh", "babi", "hello world", "saya suka makan"];
  
  for (const text of testTexts) {
    try {
      console.log(`\nTesting text: '${text}'`);
      
      const response = await fetch(
        "https://api-inference.huggingface.co/models/malaysia-ai/malaysian-sfw-classifier",
        {
          headers: {
            Authorization: `Bearer ${HF_API_KEY}`,
            "Content-Type": "application/json",
          },
          method: "POST",
          body: JSON.stringify({
            inputs: text,
            options: { wait_for_model: true }
          }),
        }
      );
      
      if (!response.ok) {
        console.log(`❌ API Error: ${response.status} ${response.statusText}`);
        const errorText = await response.text();
        console.log(`Error details: ${errorText}`);
        continue;
      }
      
      const result = await response.json();
      console.log(`✅ API Response:`, result);
      
      // Check if it's an array of classification results
      if (Array.isArray(result) && result.length > 0) {
        const scores = result[0];
        if (Array.isArray(scores)) {
          console.log(`  Classification scores:`, scores);
          const maxScore = Math.max(...scores.map(s => s.score));
          const maxLabel = scores.find(s => s.score === maxScore);
          console.log(`  Predicted: ${maxLabel.label} (${(maxScore * 100).toFixed(2)}%)`);
        }
      }
      
    } catch (error) {
      console.log(`❌ Error testing '${text}': ${error.message}`);
    }
  }
  
  return true;
}

async function testXLMRoberta() {
  console.log("\n" + "=" * 60);
  console.log("TESTING XLM-ROBERTA MULTILINGUAL TOXIC CLASSIFIER VIA API");
  console.log("=" * 60);
  
  const HF_API_KEY = process.env.HUGGINGFACE_API_KEY;
  if (!HF_API_KEY) {
    console.log("❌ No Hugging Face API key found");
    return false;
  }
  
  const testTexts = ["bodoh", "babi", "hello world", "you are stupid"];
  
  for (const text of testTexts) {
    try {
      console.log(`\nTesting text: '${text}'`);
      
      const response = await fetch(
        "https://api-inference.huggingface.co/models/unitary/multilingual-toxic-xlm-roberta",
        {
          headers: {
            Authorization: `Bearer ${HF_API_KEY}`,
            "Content-Type": "application/json",
          },
          method: "POST",
          body: JSON.stringify({
            inputs: text,
            options: { wait_for_model: true }
          }),
        }
      );
      
      if (!response.ok) {
        console.log(`❌ API Error: ${response.status} ${response.statusText}`);
        const errorText = await response.text();
        console.log(`Error details: ${errorText}`);
        continue;
      }
      
      const result = await response.json();
      console.log(`✅ API Response:`, result);
      
      // Check if it's an array of classification results
      if (Array.isArray(result) && result.length > 0) {
        const scores = result[0];
        if (Array.isArray(scores)) {
          console.log(`  Classification scores:`, scores);
          const toxicScore = scores.find(s => s.label === 'TOXIC');
          if (toxicScore) {
            console.log(`  Toxic probability: ${(toxicScore.score * 100).toFixed(2)}%`);
          }
        }
      }
      
    } catch (error) {
      console.log(`❌ Error testing '${text}': ${error.message}`);
    }
  }
  
  return true;
}

async function main() {
  console.log("ML MODEL ACCESS TESTING VIA HUGGING FACE API");
  console.log("=" * 60);
  
  // Check environment
  const HF_API_KEY = process.env.HUGGINGFACE_API_KEY;
  if (HF_API_KEY) {
    console.log(`✅ HF API Key found: ${HF_API_KEY.substring(0, 10)}...`);
  } else {
    console.log("⚠️  No HF API key found in environment");
    console.log("Please set HUGGINGFACE_API_KEY environment variable");
    return false;
  }
  
  // Test models
  const results = {};
  results.malaysian_sfw = await testMalaysianSFWClassifier();
  results.xlm_roberta = await testXLMRoberta();
  
  // Summary
  console.log("\n" + "=" * 60);
  console.log("SUMMARY");
  console.log("=" * 60);
  
  for (const [model, success] of Object.entries(results)) {
    const status = success ? "✅ ACCESSIBLE" : "❌ FAILED";
    console.log(`${model}: ${status}`);
  }
  
  const allSuccess = Object.values(results).every(Boolean);
  console.log(`\nOverall result: ${allSuccess ? "✅ SUCCESS" : "❌ SOME FAILURES"}`);
  
  return allSuccess;
}

// Run the test
main().catch(console.error);

File: test_ml_models.py
#!/usr/bin/env python3
"""
Test script to verify ML model access for cascading fallback system
"""

import os
import sys
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch

def test_malaysian_sfw_classifier():
    """Test access to Malaysian SFW Classifier"""
    print("=" * 60)
    print("TESTING MALAYSIAN SFW CLASSIFIER")
    print("=" * 60)
    
    try:
        print("Loading tokenizer...")
        tokenizer = AutoTokenizer.from_pretrained(
            'malaysia-ai/malaysian-sfw-classifier', 
            trust_remote_code=True
        )
        print("✅ Tokenizer loaded successfully!")
        
        print("Loading model...")
        model = AutoModelForSequenceClassification.from_pretrained(
            'malaysia-ai/malaysian-sfw-classifier', 
            trust_remote_code=True
        )
        print("✅ Model loaded successfully!")
        
        # Test inference with Malay profanity
        test_texts = ["bodoh", "babi", "hello world", "saya suka makan"]
        
        for text in test_texts:
            print(f"\nTesting text: '{text}'")
            inputs = tokenizer(text, return_tensors='pt', padding=True, truncation=True)
            
            with torch.no_grad():
                outputs = model(**inputs)
                logits = outputs.logits
                probabilities = torch.softmax(logits, dim=-1)
                
            print(f"  Logits: {logits}")
            print(f"  Probabilities: {probabilities}")
            print(f"  Predicted class: {torch.argmax(probabilities, dim=-1).item()}")
            
        return True
        
    except Exception as e:
        print(f"❌ Error accessing Malaysian SFW Classifier: {e}")
        print(f"Error type: {type(e).__name__}")
        return False

def test_xlm_roberta():
    """Test access to XLM-RoBERTa multilingual toxic classifier"""
    print("\n" + "=" * 60)
    print("TESTING XLM-ROBERTA MULTILINGUAL TOXIC CLASSIFIER")
    print("=" * 60)
    
    try:
        print("Loading tokenizer...")
        tokenizer = AutoTokenizer.from_pretrained('unitary/multilingual-toxic-xlm-roberta')
        print("✅ Tokenizer loaded successfully!")
        
        print("Loading model...")
        model = AutoModelForSequenceClassification.from_pretrained('unitary/multilingual-toxic-xlm-roberta')
        print("✅ Model loaded successfully!")
        
        # Test inference
        test_texts = ["bodoh", "babi", "hello world", "you are stupid"]
        
        for text in test_texts:
            print(f"\nTesting text: '{text}'")
            inputs = tokenizer(text, return_tensors='pt', padding=True, truncation=True)
            
            with torch.no_grad():
                outputs = model(**inputs)
                logits = outputs.logits
                probabilities = torch.softmax(logits, dim=-1)
                
            print(f"  Logits: {logits}")
            print(f"  Probabilities: {probabilities}")
            print(f"  Predicted class: {torch.argmax(probabilities, dim=-1).item()}")
            
        return True
        
    except Exception as e:
        print(f"❌ Error accessing XLM-RoBERTa: {e}")
        print(f"Error type: {type(e).__name__}")
        return False

def main():
    print("ML MODEL ACCESS TESTING")
    print("=" * 60)
    
    # Check environment
    hf_token = os.getenv('HUGGING_FACE_API_KEY')
    if hf_token:
        print(f"✅ HF Token found: {hf_token[:10]}...")
    else:
        print("⚠️  No HF token found in environment")
    
    print(f"Python version: {sys.version}")
    print(f"PyTorch version: {torch.__version__}")
    
    # Test models
    results = {}
    results['malaysian_sfw'] = test_malaysian_sfw_classifier()
    results['xlm_roberta'] = test_xlm_roberta()
    
    # Summary
    print("\n" + "=" * 60)
    print("SUMMARY")
    print("=" * 60)
    for model, success in results.items():
        status = "✅ ACCESSIBLE" if success else "❌ FAILED"
        print(f"{model}: {status}")
    
    return all(results.values())

if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)
